### DISCUSSION TITLE
RFC 0007: opening discussion (adding link overlay)

### DISCUSSION THREAD

**Discussion by blelump** [the-human-colossus-foundation/oca-spec#79] [General]
The [RFC 0007](https://github.com/the-human-colossus-foundation/oca-spec/blob/master/RFCs/0007-link-overlay.md) proposes `Link Overlay` as an enhancement to the `Attribute Mapping Overlay` when stronger binding is required. Here's the discussion thread covering various aspects of the proposed changes.

**Comment by carlyh-micb** [comment]
This looks good to me. Can I have more than one link overlay in a schema bundle? I could link to one or more different OCA Bundles? I assume reference can be both by refs or refn?

"In addition to the capture_base, type, and language attributes (see [Common attributes](https://github.com/the-human-colossus-foundation/oca-spec/blob/master/RFCs/0007-link-overlay.md#common-attributes)), the Entry Overlay MUST include the following attribute:"

In the example given there is no language - are languages required to be present in overlays that don't have a language requirements? It appears that the Link overlay doesn't need to be specific for language as it links attr to attr. Thus is the language attribute required?

Can this overlay support entry code linking as well or will there be a separate linking overlay for linking entry codes in one schema to entry codes in a second schema? 

@pknowl do you have an answer for: "Is linking a proper name for the underlying concept?"

**Comment by pknowl** [comment]
The currently proposed "Link Overlay" is intended to establish explicit, unidirectional correspondences between attributes in a source Bundle and attributes in a target Bundle. While the overlay performs a mapping function, the term "Link" introduces ambiguity and does not clearly describe its technical purpose.

**Rationale for Renaming**

The difference between **mapping** and **framing** highlights the issue with the current terminology:

**Mapping**:
- Involves creating explicit correspondences between two data structures, typically at the attribute level.
- Ensures compatibility and interoperability between systems.
- Operates on a technical, structural level.

**Framing**:
- Provides semantic context by connecting data to external vocabularies, ontologies, or conceptual models.
- Operates at a higher conceptual level to ensure shared understanding across domains.

The "Link Overlay" as proposed is solely focused on mapping attributes between Bundles. It does not provide semantic context or framing capabilities, making the term "link" both overly broad and technically imprecise.

**Issues with the Term "Link"**

- **Generic and Ambiguous**: "Link" is widely used across various contexts (e.g., hyperlinks, database relationships, network connections). Users may misunderstand its role within the OCA ecosystem.

- **Misaligned Expectations**: "Link" often suggests bidirectional or general connections, whereas the overlay establishes explicit, unidirectional mappings at the attribute level.

- **Undervalues Functionality**: The term does not emphasize the overlay’s critical role in aligning and mapping structured data between Bundles.

**Proposed Solution**

Rename the "Link Overlay" to "Mapping Overlay" to better align with its function and purpose. This change would:

- Clearly convey the overlay's role in creating attribute-level mappings.
- Differentiate it from framing overlays, which provide semantic context.
- Establish a meaningful synergy between Mapping Overlays (structural alignment) and Framing Overlays (semantic alignment).

**Benefits of Renaming**

- **Clarity**: The name "Mapping Overlay" directly communicates the overlay's functionality, reducing potential confusion.

- **Consistency**: Aligns with existing terminology in data management, where mapping is a well-understood concept.

- **Synergy**: Differentiates mapping and framing overlays while emphasizing their complementary roles within OCA.

**Conclusion**

The proposed "Link Overlay" is functionally a "Mapping Overlay". Renaming it will improve clarity, precision, and alignment with established concepts in data management and the OCA ecosystem.

**Discussion Points**
Do we agree that "Mapping Overlay" better describes the overlay’s purpose and functionality?

**Comment by pknowl** [comment]
We've identified some critical gaps in the Mapping Overlay that need to be addressed for it to effectively support AI-driven workflows and relational data operations. Currently, the overlay allows for merging two given OCA data records but lacks support for operating over sets, which is essential for our use cases.

Key Required Enhancements:

1. **Support for Two Bundles** – The Mapping Overlay should explicitly reference two bundles rather than assuming a one-way mapping.

2. **Handling of Optional Attributes** – We need a way to merge columns where attributes may be missing from one side of the mapping.

3. **Description of Relationships Between Bundles** – The Linking Identifier needs to be explicitly defined, with support for composite keys (i.e., multiple attributes forming a compound identifier).

4. **Support for Data Transformation (Future Enhancement)** – While not immediately critical, enabling transformation logic (e.g., normalization, concatenation, type conversion) would be a valuable addition down the line.

Additional Concern:

- **Missing Identifier Concept in Capture Base** – Ryan pointed out that we overlooked the identifier concept in the capture base, likely because our initial focus was on credentials rather than relational data structures. This is an oversight we should correct.

The attached screenshots illustrate the issue in the context of merging datasets with both identical and divergent structures. Let’s discuss the best approach to extend the Mapping Overlay and capture base accordingly.

**Comment by pknowl** [comment]
<img width="734" alt="Screenshot 2025-01-28 at 21 10 25" src="https://github.com/user-attachments/assets/59157499-23cc-4b57-a4be-1f60f078eec6" />
<img width="670" alt="Screenshot 2025-01-28 at 21 10 43" src="https://github.com/user-attachments/assets/e4c768c4-50dc-40f4-953a-d8ac583b577a" />
<img width="1032" alt="Screenshot 2025-01-28 at 21 11 26" src="https://github.com/user-attachments/assets/1cd6ff82-0036-40d3-986d-4102514aa7a5" />
